File: leetcode-51_N-Queens.go
/**
 * N皇后问题
 */
package main

import "fmt"

func main() {
    fmt.Println(solveNQueens(5))
}

var ret [][]string

func solveNQueens(n int) [][]string {
    nQueens := [][]string{}
    for i := 0; i < n; i++ { //初始化一个二维数组
        nQueens = append(nQueens, make([]string, n))
    }

    dfs(nQueens, 0, n)
    return ret
}

//递归求N皇后解法
func dfs(nQueens [][]string, row, n int) {
    if row == n { //到达最后一行
        //nQueens数组即一个解, 这里拼成字符串是题目要求
        tmp := []string{}
        for i := 0; i < n; i++ {
            str := ""
            for j := 0; j < n; j++ {
                if nQueens[i][j] != "" {
                    str += nQueens[i][j]
                } else {
                    str += "."
                }
            }
            tmp = append(tmp, str)
        }
        ret = append(ret, tmp)
        return
    }

    //在新行内逐个列尝试
    for col := 0; col < n; col++ {
        if isValid(nQueens, row, col, n) {
            nQueens[row][col] = "Q" //放置皇后
            dfs(nQueens, row + 1, n) //进入下一行
            nQueens[row][col] = "." //取出该位置的皇后, 尝试下一个col
        }
    }
}

//判断某个坐标能否放置皇后
func isValid(nQueens [][]string, row, col, n int) bool {
    //判断已有皇后的行, 当前列是否有冲突
    for i := 0; i < row; i++ {
        if nQueens[i][col] == "Q" {
            return false
        }
    }

    //判断左上方斜线上是否有皇后
    for i, j := row - 1, col - 1; i >= 0 && j >= 0; i, j = i - 1, j - 1 {
        if nQueens[i][j] == "Q" {
            return false
        }
    }

    //判断右上方斜线上是否有皇后
    for i, j := row - 1, col + 1; i >= 0 && j < n; i, j = i - 1, j + 1 {
        if nQueens[i][j] == "Q" {
            return false
        }
    }

    return true
}

File: leetcode-51_n-queens.java
/**
 * N 皇后问题
 */
class Solution {
    // 返回结果
    private List<List<String>> ret = new ArrayList<>();

    // N 皇后问题
    public List<List<String>> solveNQueens(int n) {
        char[][] chessboard = new char[n][n];
        dfs(chessboard, 0, n);
        return ret;
    }

    public void dfs(char[][] chessboard, int row, int n) {
        // 已经超出棋盘最后一行(即n个棋子已摆放完毕)
        if (row == n) {
            // 打印当前摆放结果
            ret.add(printChessboard(chessboard));
            return;
        }

        // 一行内逐个列尝试
        for (int col = 0; col < n; col++) {
            if (isValid(chessboard, row, col, n)) {
                chessboard[row][col] = 'Q';
                // 当前行放下棋子后，继续尝试下一行
                dfs(chessboard, row + 1, n);
                // 下一行尝试完后，回到当前行，收起刚才放下的棋子，for 循环尝试当前行的下一列
                chessboard[row][col] = '.';
            }
        }
    }

    public boolean isValid(char[][] chessboard, int row, int col, int n) {
        // 同一列是否已有棋子
        for (int i = 0; i < row; i++) {
            if (chessboard[i][col] == 'Q') {
                return false;
            }
        }

        // 左上方斜线上是否有棋子
        for (int i = row - 1, j = col - 1; i >= 0 && j >= 0; i = i - 1, j = j - 1) {
            if (chessboard[i][j] == 'Q') {
                return false;
            }
        }

        // 右上方斜线上是否有棋子
        for (int i = row - 1, j = col + 1; i >= 0 && j < n; i = i - 1, j = j + 1) {
            if (chessboard[i][j] == 'Q') {
                return false;
            }
        }

        return true;
    }

    // 打印棋子摆放结果
    public List<String> printChessboard(char[][] chessboard) {
        List<String> cb = new ArrayList<>();
        for (int i = 0; i < chessboard.length; i++) {
            for (int j = 0; j < chessboard[0].length; j++) {
                if (chessboard[i][j] == '\0') {
                    chessboard[i][j] = '.';
                }
            }
            cb.add(new String(chessboard[i]));
        }
        return cb;
    }
}
